### PR TITLE
Support inherited fields in parameter validation and navigation

### PR DESCRIPTION
**Pull request by young1lin**
## Summary

Resolves issue #50 by implementing comprehensive support for inherited Java fields across the extension's parameter validation and XML-to-Java navigation features. Previously, only fields declared directly on a class were recognized; now the extension walks the `extends` chain to resolve fields inherited from superclasses.

## Key Changes

- **New Module: `javaFieldHierarchy.ts`** - Core utility for walking Java class hierarchies
  - `getClassFieldsWithInheritance()`: Collects own and inherited fields across the entire superclass chain
  - `findFieldInHierarchy()`: Locates a specific field by name, stopping at the declaring class
  - Handles edge cases: cyclic inheritance, missing workspace files, built-in types (java.lang.*), and depth limits (max 10 levels)
  - Subclass fields properly shadow same-named superclass fields

- **Parser Enhancement: `javaFieldParser.ts` & `javaTreeSitterParser.ts`**
  - New `extractSuperclassName()` function with two-tier fallback (WASM → Regex)
  - New `extractSuperclassNameFromAST()` for tree-sitter-based extraction
  - Strips generic type arguments (e.g., `BaseEntity<Long>` → `BaseEntity`)
  - Handles fully-qualified superclass names and multi-line class declarations

- **ParameterValidator Updates**
  - Now uses `getClassFieldsWithInheritance()` instead of direct field extraction
  - Implements smart cache invalidation: editing a superclass invalidates cached entries of all its subclasses
  - Tracks class dependencies via `classDependents` map to efficiently invalidate affected entries
  - Maintains backward compatibility with existing parameter validation logic

- **Navigation Provider Updates**
  - `XmlParameterDefinitionProvider`: Uses `findFieldInHierarchy()` to navigate to inherited fields
  - `XmlResultMapPropertyDefinitionProvider`: Uses `findFieldInHierarchy()` for property navigation
  - Both providers now correctly land on the file that actually declares the field, even if inherited

- **Comprehensive Test Coverage**
  - Unit tests for `javaFieldHierarchy` with real fixture files (3-level hierarchy: TaskQuery → AuditEntity → BaseEntity<Long>)
  - Edge case tests: field shadowing, missing superclasses, unresolvable types, cyclic chains, depth limits
  - Unit tests for superclass extraction (AST and regex fallback)
  - Integration tests for parameter validation with inherited fields
  - Integration tests for XML-to-Java navigation with inherited fields
  - Cache invalidation tests for ParameterValidator

- **Test Fixtures**
  - Added 3-level class hierarchy: `TaskQuery` → `AuditEntity` → `BaseEntity<Long>`
  - Added `TaskMapper.xml` and `TaskMapper.java` demonstrating inherited field usage
  - Fixtures validate both parameter validation and navigation scenarios

## Implementation Details

- **Hierarchy Walking**: Uses async generator pattern for memory efficiency and early termination
- **Cycle Detection**: Maintains visited set to prevent infinite loops in pathological inheritance chains
- **Depth Limiting**: Caps hierarchy walk at 10 levels to prevent performance degradation
- **Type Resolution**: Leverages existing `resolveFullyQualifiedType()` to convert simple names to FQNs
- **Cache Strategy**: Field cache now tracks class dependencies, enabling targeted invalidation when superclasses change
- **Backward Compatibility**: All changes are additive; existing APIs remain unchanged

https://claude.ai/code/session_018zwgiFf2M4SCLsqiWnRrYD

---

> [!NOTE]
> **Medium Risk**
> Touches core XML parameter validation, definition navigation, and type resolution precedence; incorrect hierarchy or cache invalidation could cause false diagnostics or wrong jump targets, though coverage is extensive.
> 
> **Overview**
> Fixes **issue #50** by treating **inherited** Java fields like declared ones for MyBatis XML parameter checks and go-to-definition.
> 
> A new **`javaFieldHierarchy`** helper walks the `extends` chain (with depth/cycle guards and **subclass shadowing**), backed by **`extractSuperclassName`** and **class-scoped** field parsing so multi-type source files do not leak the wrong members. **`ParameterValidator`** now builds valid parameter names from that hierarchy and keeps an **inheritance-aware field cache** (superclass edits and LRU eviction invalidate dependents; saving Java re-runs open XML diagnostics). **`XmlParameterDefinitionProvider`** and **`XmlResultMapPropertyDefinitionProvider`** jump to the **declaring** superclass file instead of only the parameter/result type class.
> 
> **`javaTypeResolver`** is reordered to match Java precedence (explicit import → same package → wildcard → Java LS). **`esbuild`** forces the **CommonJS** `web-tree-sitter` entry so bundled parsing keeps working. Fixtures, Java integration samples, and unit/integration tests cover multi-level hierarchies, shadowing, and cache behavior; some navigation integration suites are re-enabled or fixed for workspace-scoped runs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 670d14922e43ec758cde8d5f6af8d5626c2020b4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/utils/javaFieldHierarchy.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        visited.add(current);
+
+        const file = await findJavaClassFile(current);
+        if (!file) {
+            break;
+        }
+
+        yield { className: current, filePath: file.fsPath };
+
+        current = await resolveSuperclass(file.fsPath);
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/utils/javaFieldHierarchy.ts]: Fixed in 17f7fbb. Superclass extraction is now anchored to the class being walked: the walker passes the visited FQN's simple name down, the AST tier only inspects the matching `class_declaration`, and the regex fallback anchors on the class name and strips line/block comments first. Added regression tests for the secondary-class and comment false-positive scenarios.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/javaFieldHierarchy.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+/**
+ * Resolve the fully-qualified superclass name of the class declared in a file
+ */
+async function resolveSuperclass(filePath: string): Promise<string | null> {
+    try {
+        const superclassName = await extractSuperclassName(filePath);
+        if (!superclassName || isBuiltInType(superclassName)) {
+            return null;
+        }
+        const fullyQualified = await resolveFullyQualifiedType(filePath, superclassName);
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/utils/javaFieldHierarchy.ts:138]: Fixed in 17f7fbb. `resolveFullyQualifiedType` now has a wildcard-import tier at the lowest precedence (explicit import > same package > wildcard, matching Java semantics): each `import pkg.*;` package is probed for a matching workspace source file. Static imports are ignored. Covered by new unit tests including the precedence case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/javaFieldHierarchy.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    for await (const cls of walkClassHierarchy(className)) {
+        classChain.push(cls.className);
+
+        const ownFields = await extractJavaFields(cls.filePath);
+        for (const field of ownFields) {
+            if (!seenFieldNames.has(field.name)) {
+                seenFieldNames.add(field.name);
+                fields.push({ field, filePath: cls.filePath, className: cls.className });
+            }
+        }
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/utils/javaFieldHierarchy.ts:94]: Fixed in 1a0d276. `extractJavaFields` now takes an optional class name, threaded through all three tiers: the AST tier returns only the matching class's direct members (nested types excluded), the LS tier filters by class symbol name, and the regex tier tracks which top-level class body it's in. The hierarchy walker passes the visited FQN's simple name for every level, so sibling/helper classes can no longer contribute or shadow fields. Callers without a class name keep the legacy whole-file behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/navigator/parsers/javaFieldParser.ts [hunk 4]
```diff
@@ -84,8 +85,53 @@ export async function findJavaFieldPosition(
...
+    // The optional generic group keeps bounded type parameters like
+    // "class Child<T extends Number> extends Base" from capturing "Number".
+    // Anchoring on the class name keeps other types in the same file from
+    // being mistaken for the class's superclass.
+    const classToken = className ? escapeRegex(className) : '\\w+';
+    const superclassRegex = new RegExp(
+        `\\bclass\\s+${classToken}(?:\\s*<[^>]*>)?\\s+extends\\s+([\\w.]+)`
+    );
+    const match = withoutComments.match(superclassRegex);
+    return match ? match[1] : null;
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/navigator/parsers/javaFieldParser.ts:136]: Fixed in 1a0d276. The generic group is now `(?:\s*<(?:[^<>]|<[^<>]*>)*>)?`, which handles one level of nesting (`class Foo<T extends Comparable<String>> extends Base` → `Base`) while still not capturing bounded type parameters. Two-plus levels of nesting (e.g. `T extends List<Map<K,V>>`) still fall back to null in this tier — the AST tier handles those, and regex is only reached when tree-sitter is unavailable.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/navigator/diagnostics/ParameterValidator.ts [hunk 5]
```diff
@@ -105,15 +106,15 @@ export class ParameterValidator {
         // Watch Java files at filesystem level for external changes (git checkout, etc.)
         const javaFsWatcher = vscode.workspace.createFileSystemWatcher('**/*.java');
         javaFsWatcher.onDidChange(() => {
-            this.fieldCache.clear();
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/navigator/diagnostics/ParameterValidator.ts:108]: Addressed in 1a0d276, with one deliberate scoping note. The save handler now calls `revalidateOpenXmlDocuments()` explicitly instead of relying on the `**/*.java` fs watcher (which also fires on save and already triggered revalidation — the stale-until-XML-edit scenario only occurred if the watcher missed the event). The `onDidChangeTextDocument` (dirty buffer) path intentionally does not revalidate: all field parsers read from disk, so revalidating on unsaved keystrokes would re-produce identical diagnostics while adding churn. Diagnostics refresh at save, which is when the on-disk truth actually changes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/navigator/diagnostics/ParameterValidator.ts [hunk 6]
```diff
@@ -393,26 +398,27 @@ export class ParameterValidator {
...
-            const fields = await extractJavaFields(file.fsPath);
-            const fieldNames = fields.map(f => f.name);
+            // 4. Record that editing any class in the chain invalidates this entry
+            for (const chainClass of classChain) {
+                let dependents = this.classDependents.get(chainClass);
+                if (!dependents) {
+                    dependents = new Set();
+                    this.classDependents.set(chainClass, dependents);
+                }
+                dependents.add(className);
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/navigator/diagnostics/ParameterValidator.ts:426]: Fixed in 4b2b66b. `LRUCache` now takes an optional `onEvict` callback (fired only for capacity evictions — explicit delete/clear are the owner's own actions), and the validator keeps a forward index of cache key → inheritance chain. Whenever an entry leaves the cache (invalidation, silent eviction, or clear), `unregisterHierarchy` removes it from every dependent set and prunes empty entries, so `classDependents` stays bounded by the live cache contents. Covered by new LRUCache callback tests and a validator test that fills the cache past capacity and asserts the bookkeeping is pruned.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/navigator/parsers/javaFieldParser.ts [hunk 7]
```diff
@@ -108,7 +163,7 @@ function extractJavaFieldsRegex(content: string): JavaField[] {
             inClassBody = true;
         }
 
-        if (!inClassBody || braceLevel !== 1) {
+        if (!inClassBody || braceLevel !== 1 || !inTargetClass) {
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/navigator/parsers/javaFieldParser.ts:171]: Fixed in 2e2d419. The regex scoper now only updates the in-target-class flag for top-level declarations (`braceLevel === 0`); nested type declarations no longer clear it. Nested types' own fields stay excluded by the existing brace-depth check, and outer-class fields declared after a nested type are collected again. Regression test added (outer field → nested class → outer field).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/navigator/diagnostics/ParameterValidator.ts [hunk 8]
```diff
@@ -393,26 +402,28 @@ export class ParameterValidator {
...
+            // 4. Record that editing any class in the chain invalidates this entry
+            this.dependencyChains.set(className, classChain);
+            for (const chainClass of classChain) {
+                let dependents = this.classDependents.get(chainClass);
+                if (!dependents) {
+                    dependents = new Set();
+                    this.classDependents.set(chainClass, dependents);
+                }
+                dependents.add(className);
+            }
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/navigator/diagnostics/ParameterValidator.ts:427]: Fixed in e3b3166. `getClassFields` now calls `unregisterHierarchy(className)` before re-registering, so a repopulating walk always replaces the previous chain's links instead of stacking on them. Test added for the overlapping-miss scenario (two concurrent misses whose walks return different chains — the superseded parent link is dropped and the forward index holds the latest chain).

One scope note: the "validation may stay wrong until a full cache clear" outcome was already unreachable — walks can only differ if the on-disk hierarchy changed between them, and any such change fires the `**/*.java` watcher, which clears the whole cache and revalidates. The residue this fixes was conservative over-invalidation, not stale results.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/javaFieldHierarchy.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+ * Resolve the fully-qualified superclass name of a specific class declared in a file.
+ * The class name anchors extraction so other types in the same compilation unit
+ * cannot be mistaken for the class's superclass.
+ */
+async function resolveSuperclass(filePath: string, className: string): Promise<string | null> {
+    try {
+        const superclassName = await extractSuperclassName(filePath, className);
+        if (!superclassName || isBuiltInType(superclassName)) {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/utils/javaFieldHierarchy.ts:133]: Fixed in 038561e. Good catch — and it's even stronger than stated: java.lang's String/Integer/etc. are `final`, so a compiling `extends Integer` can *only* be a user class shadowing the name. The pre-resolution short-name check is removed; only the resolved FQN is tested against java.lang built-ins (which still stops `extends Object` once resolution lands on `java.lang.Object`). Tests added for both directions: the shadowing user class is walked, the true java.lang resolution stops the walk.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/javaTypeResolver.ts [hunk 10]
```diff
@@ -68,13 +68,20 @@ export async function resolveFullyQualifiedType(
...
 
     // Same-package fallback
-    const result = await resolveFromSamePackage(content, simpleTypeName);
-    if (result) {
-        console.log(`[javaTypeResolver] Same-package resolved ${simpleTypeName} to ${result}`);
+    const fromSamePackage = await resolveFromSamePackage(content, simpleTypeName);
+    if (fromSamePackage) {
+        console.log(`[javaTypeResolver] Same-package resolved ${simpleTypeName} to ${fromSamePackage}`);
+        return fromSamePackage;
+    }
```
[the remark this answers is not in the document]
Comment (young1lin) [reply src/utils/javaTypeResolver.ts:67]: Fixed in 038561e. `resolveFullyQualifiedType` now follows Java's simple-name precedence end to end: explicit imports → same package → wildcard imports → Java LS last. The LS tier only fires when nothing in the workspace can satisfy the name, which is exactly its purpose (classpath types without workspace sources). Added a test asserting a same-package file wins over an LS match from another package — and that the LS isn't even consulted in that case. This also corrects the same latent ordering for the pre-existing `parameterType`/auto-map resolution paths that share this utility.

---
_Generated by [Claude Code](https://claude.ai/code)_
